include/arch/riscv/syscall.h

```diff
@@ -0,0 +1,164 @@
+	register uint32_t a2 __asm__ ("a2") = arg3;
+	register uint32_t a3 __asm__ ("a3") = arg4;
+	register uint32_t a4 __asm__ ("a4") = arg5;
+	register uint32_t a5 __asm__ ("a5") = arg6;
+	register uint32_t a7 __asm__ ("a7") = call_id;
+
+	__asm__ volatile ("ecall"
+			  : "+r" (a0)
+			  : "r" (a1), "r" (a2), "r" (a3), "r" (a4), "r" (a5),
+			  "r" (a7)
+			  : "memory");
+	return a0;
+}
+
+static inline uintptr_t arch_syscall_invoke5(uintptr_t arg1, uintptr_t arg2,
+					     uintptr_t arg3, uintptr_t arg4,
+					     uintptr_t arg5,
+					     uintptr_t call_id)
+{
+	register uint32_t a0 __asm__ ("a0") = arg1;
+	register uint32_t a1 __asm__ ("a1") = arg2;
+	register uint32_t a2 __asm__ ("a2") = arg3;
+	register uint32_t a3 __asm__ ("a3") = arg4;
+	register uint32_t a4 __asm__ ("a4") = arg5;
+	register uint32_t a7 __asm__ ("a7") = call_id;
+
+	__asm__ volatile ("ecall"
+			  : "+r" (a0)
+			  : "r" (a1), "r" (a2), "r" (a3), "r" (a4), "r" (a7)
+			  : "memory");
+	return a0;
+}
+
+static inline uintptr_t arch_syscall_invoke4(uintptr_t arg1, uintptr_t arg2,
+					     uintptr_t arg3, uintptr_t arg4,
+					     uintptr_t call_id)
+{
+	register uint32_t a0 __asm__ ("a0") = arg1;
+	register uint32_t a1 __asm__ ("a1") = arg2;
+	register uint32_t a2 __asm__ ("a2") = arg3;
+	register uint32_t a3 __asm__ ("a3") = arg4;
+	register uint32_t a7 __asm__ ("a7") = call_id;
+
+	__asm__ volatile ("ecall"
+			  : "+r" (a0)
+			  : "r" (a1), "r" (a2), "r" (a3), "r" (a7)
+			  : "memory");
+	return a0;
+}
+
+static inline uintptr_t arch_syscall_invoke3(uintptr_t arg1, uintptr_t arg2,
+					     uintptr_t arg3,
+					     uintptr_t call_id)
+{
+	register uint32_t a0 __asm__ ("a0") = arg1;
+	register uint32_t a1 __asm__ ("a1") = arg2;
+	register uint32_t a2 __asm__ ("a2") = arg3;
+	register uint32_t a7 __asm__ ("a7") = call_id;
+
+	__asm__ volatile ("ecall"
+			  : "+r" (a0)
+			  : "r" (a1), "r" (a2), "r" (a7)
+			  : "memory");
+	return a0;
+}
+
+static inline uintptr_t arch_syscall_invoke2(uintptr_t arg1, uintptr_t arg2,
+					     uintptr_t call_id)
+{
+	register uint32_t a0 __asm__ ("a0") = arg1;
+	register uint32_t a1 __asm__ ("a1") = arg2;
+	register uint32_t a7 __asm__ ("a7") = call_id;
+
+	__asm__ volatile ("ecall"
+			  : "+r" (a0)
+			  : "r" (a1), "r" (a7)
+			  : "memory");
+	return a0;
+}
+
+static inline uintptr_t arch_syscall_invoke1(uintptr_t arg1, uintptr_t call_id)
+{
+	register uint32_t a0 __asm__ ("a0") = arg1;
+	register uint32_t a7 __asm__ ("a7") = call_id;
+
+	__asm__ volatile ("ecall"
+			  : "+r" (a0)
+			  : "r" (a7)
+			  : "memory");
+	return a0;
+}
+
+static inline uintptr_t arch_syscall_invoke0(uintptr_t call_id)
+{
+	register uint32_t a0 __asm__ ("a0");
+	register uint32_t a7 __asm__ ("a7") = call_id;
+
+	__asm__ volatile ("ecall"
+			  : "+r" (a0)
+			  : "r" (a7)
+			  : "memory");
+	return a0;
+}
+
+static inline bool arch_is_user_context(void)
+{
+	/* Defined in arch/riscv/core/thread.c */
+	extern ulong_t is_user_mode;
```

Comment:
does this actually work when called from user mode? the thread wouldn't have access to this global variable's memory unless I'm missing something.

another concern is if this is correct at all times -- what do you get if a user thread was running and you take an interrupt or exception? does the irq/exception entry/exit paths also update this?

is there some HW register state you can check instead? that would make this much simpler

Reply:
> does this actually work when called from user mode? the thread wouldn't have access to this global variable's memory unless I'm missing something.
All thread (user/kernel) have access to this global variable. User have read only access. There are 2 way to implement this function, I tried both and the second one is to enter in fault by trying to do an illegal instruction for an user (try to read mstatus register for example), but trigger fault every time you want to know if you are in userspace slow system. I choose to spend one PMP slot to up performance (this is litteraly the function of PMP).
> 
> another concern is if this is correct at all times -- what do you get if a user thread was running and you take an interrupt or exception? does the irq/exception entry/exit paths also update this?
I got the right value anytime. irq/exception entry/exit update this.
> 
> is there some HW register state you can check instead? that would make this much simpler
Unfortunately not, this is one of the RISCV limitation which lot of people are complaining.

Reply:
I can confirm that this is indeed a limitation on RISC-V. There's no CPU register that indicates the current privilege level; only the privilege level of the code that triggered the current interrupt. I implemented `arch_is_user_context` in the other PR the way @amergnat described (attempting an privileged operation and seeing if a fault occurred), but it's a lot slower and hackier.